rust_src/src/editfns.rs

```diff
@@ -209,12 +211,68 @@ pub fn insert_byte(byte: EmacsInt, count: LispObject, inherit: LispObject) -> Li
         } else {
             byte
         };
-    unsafe {
-        LispObject::from_raw(Finsert_char(
-            LispObject::from_natnum(toinsert).to_raw(),
-            count.to_raw(),
-            inherit.to_raw(),
-        ))
+    insert_char(toinsert as Codepoint, count, inherit);
+}
+
+/// Insert COUNT copies of CHARACTER.
+/// Interactively, prompt for CHARACTER.  You can specify CHARACTER in one
+/// of these ways:
+///
+///  - As its Unicode character name, e.g. \"LATIN SMALL LETTER A\".
+///    Completion is available; if you type a substring of the name
+///    preceded by an asterisk `*', Emacs shows all names which include
+///    that substring, not necessarily at the beginning of the name.
+///
+///  - As a hexadecimal code point, e.g. 263A.  Note that code points in
+///    Emacs are equivalent to Unicode up to 10FFFF (which is the limit of
+///    the Unicode code space).
+///
+///  - As a code point with a radix specified with #, e.g. #o21430
+///    (octal), #x2318 (hex), or #10r8984 (decimal).
+///
+/// If called interactively, COUNT is given by the prefix argument.  If
+/// omitted or nil, it defaults to 1.
+///
+/// Inserting the character(s) relocates point and before-insertion
+/// markers in the same ways as the function `insert'.
+///
+/// The optional third argument INHERIT, if non-nil, says to inherit text
+/// properties from adjoining text, if those properties are sticky.  If
+/// called interactively, INHERIT is t.
+#[lisp_fn(min = "1", intspec = "(list (read-char-by-name \"Insert character (Unicode name or hex): \") (prefix-numeric-value current-prefix-arg) t))")]
+pub fn insert_char(character: Codepoint, count: Option<EmacsInt>, inherit: bool) {
+    let count = count.unwrap_or(1);
+    if count <= 0 {
+        return;
+    }
+
+    //let buf = ThreadState::current_buffer();
+    //let multibyte: bool = LispObject::from_raw(buf.enable_multibyte_characters).is_not_nil().into();
+    let mut str = [0_u8; MAX_MULTIBYTE_LENGTH];
+    let len = write_codepoint(&mut str[..], character);
+
+    if BUF_BYTES_MAX / (len as isize) < (count as isize) {
+        unsafe { buffer_overflow() };
+    }
+    let mut n: ptrdiff_t = (count * (len as EmacsInt)) as ptrdiff_t;
+    let mut buffer = [0_i8; 4000];
```

Comment:
Where did 4k come from?

Reply:
It's a magic constant from the original code. The original also uses `sizeof` to avoid repeating it the next two times, but that's not so useful for us.